### PR TITLE
[KYUUBI #3936] Parse trino http request header

### LINKED ISSUE #3936: [Subtask] Parse Trino http header
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues) and found no similar issues.

### Describe the subtask

TODO

### Parent issue

#3901

### Are you willing to submit PR?

- [x] Yes. I can submit a PR independently to improve.
- [ ] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by iodone**
close #3936 
### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoContext.scala [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+import scala.collection.JavaConverters._
+
+import io.trino.client.{ClientSelectedRole, QueryResults}
+import io.trino.client.ProtocolHeaders.TRINO_HEADERS
+
+// TODO: Request and Respone context, to be enriched
+case class TrinoContext(
+    user: String,
+    timeZone: String,
+    transactionId: String,
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoContext.scala:33]: We do not support transaction, should throw exception if it's defined in header.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoContext.scala [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    clientCapabilities: String,
+    source: Option[String] = None,
+    catalog: Option[String] = None,
+    schema: Option[String] = None,
+    path: Option[String] = None,
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoContext.scala:38]: ditto

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoContext.scala [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    language: Option[String] = None,
+    traceToken: Option[String] = None,
+    clientInfo: Option[String] = None,
+    clientTags: Set[String] = Set.empty,
+    session: Map[String, String] = Map.empty,
+    role: Map[String, ClientSelectedRole] = Map.empty,
+    preparedStatement: Map[String, String] = Map.empty,
+    resourceEstimate: Map[String, String] = Map.empty,
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoContext.scala:46]: ditto

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoContext.scala [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    source: Option[String] = None,
+    catalog: Option[String] = None,
+    schema: Option[String] = None,
+    path: Option[String] = None,
+    language: Option[String] = None,
+    traceToken: Option[String] = None,
+    clientInfo: Option[String] = None,
+    clientTags: Set[String] = Set.empty,
+    session: Map[String, String] = Map.empty,
+    role: Map[String, ClientSelectedRole] = Map.empty,
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoContext.scala:44]: ditto

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoContext.scala [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    preparedStatement: Map[String, String] = Map.empty,
+    resourceEstimate: Map[String, String] = Map.empty,
+    extraCredentials: Map[String, String] = Map.empty) {}
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoContext.scala:47]: ditto

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/v1/StatementResource.scala [hunk 6]
```diff
@@ -54,8 +55,21 @@ private[v1] class StatementResource extends ApiRequestContext with Logging {
...
+      URI.create("cancel"),
+      URI.create("next"),
+      null,
+      null,
+      null,
+      null,
+      null,
+      "update_type",
+      0)
+    TrinoContext.buildTrinoResponse(qr, trinoContext)
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/v1/StatementResource.scala:72]: it still can not work after this pr. I think keeping throw a exception is better

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoContext.scala [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    headers.foldLeft(requestCtx) { case (context, (k, v)) =>
+      k match {
+        case k if TRINO_HEADERS.requestUser.equalsIgnoreCase(k) && v.nonEmpty =>
+          context.copy(user = v.head)
+        case k if TRINO_HEADERS.requestTimeZone.equalsIgnoreCase(k) =>
+          context.copy(timeZone = v.headOption)
+        case k
+            if TRINO_HEADERS.requestTransactionId.equalsIgnoreCase(k)
+              && v.headOption.exists(_ != "NONE") =>
+          throw new UnsupportedOperationException(s"$k is not currently supported")
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/TrinoContext.scala:85]: can we move all unsupported head to the end of match-case ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/api.scala [hunk 8]
```diff
@@ -35,6 +35,8 @@ private[api] trait ApiRequestContext {
   @Context
   protected var httpRequest: HttpServletRequest = _
 
+  protected var trinoContext: TrinoContext = _
```
[the remark this answers is not in the document]
Comment (iodone) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/api.scala:38]: I plan to address this in the next issue. In the next issue(#3935) , there can be a complete request-response process, which will be more helpful for the design and implementation of `session`. WDYT?

[the remark this answers is not in the document]
Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/trino/api/api.scala:38]: That's ok. but we should remove this code in this pr. It's no used and just make reviewer confused

**Review by ulysses-you — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merging to master

**Pull request closed without merging**